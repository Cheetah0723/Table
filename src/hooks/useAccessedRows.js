import { useMemo } from 'react'

export default function useAccessedRows ({
  debug, data, columns, subRowsKey,
}) {
  return useMemo(
    () => {
      if (debug) console.info('getAccessedRows')

      // Access the row's data
      const accessRow = (data, i, depth = 0) => {
        // Keep the original reference around
        const original = data

        // Process any subRows
        const subRows = data[subRowsKey]
          ? data[subRowsKey].map((d, i) => accessRow(d, i, depth + 1))
          : undefined

        const row = {
          original,
          index: i,
          subRows,
          depth,
        }

        // Create the cells and values
        row.values = {}
        columns.forEach(column => {
          row.values[column.id] = column.accessor ? column.accessor(data) : undefined
        })

        return row
      }

      // Use the resolved data
      return data.map((d, i) => accessRow(d, i))
    },
    [data, columns]
  )
}
