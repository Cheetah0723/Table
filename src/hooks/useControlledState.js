const mergeStateAndProps = (state, defaults, props) => {
  Object.keys(props).forEach(key => {
    if (typeof props[key] !== 'undefined') {
      state[key] = props[key]
    }
  })

  Object.keys(defaults).forEach(key => {
    if (typeof state[key] === 'undefined') {
      state[key] = defaults[key]
    }
  })

  return state
}

export default function useControlledState ([userState, userSetState], defaults, props) {
  const state = mergeStateAndProps(userState, defaults, props)

  const setState = (updater, type) =>
    userSetState(old => updater(mergeStateAndProps(old, defaults, props)), type)

  return [state, setState]
}
