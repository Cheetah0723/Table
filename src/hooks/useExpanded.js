import { useMemo } from 'react'
import PropTypes from 'prop-types'

import { getBy, getFirstDefined, setBy } from '../utils'
import { addActions, actions } from '../actions'
import { defaultState } from './useTableState'

defaultState.expanded = {}

addActions({
  toggleExpanded: '__toggleExpanded__',
  useExpanded: '__useExpanded__'
})

const propTypes = {
  expandedKey: PropTypes.string,
  paginateSubRows: PropTypes.bool
}

export const useExpanded = props => {
  PropTypes.checkPropTypes(propTypes, props, 'property', 'useExpanded')

  const {
    debug,
    rows,
    expandedKey = 'expanded',
    hooks,
    state: [{ expanded }, setState],
    paginateSubRows = true
  } = props

  const toggleExpandedByPath = (path, set) => {
    return setState(old => {
      const { expanded } = old
      const existing = getBy(expanded, path)
      set = getFirstDefined(set, !existing)
      return {
        ...old,
        expanded: setBy(expanded, path, set)
      }
    }, actions.toggleExpanded)
  }

  hooks.row.push(row => {
    const { path } = row
    row.toggleExpanded = set => toggleExpandedByPath(path, set)
  })

  const expandedRows = useMemo(() => {
    if (debug) console.info('getExpandedRows')

    const expandedRows = []

    // Here we do some mutation, but it's the last stage in the
    // immutable process so this is safe
    const handleRow = (row, index, depth = 0, parentPath = []) => {
      // Compute some final state for the row
      const path = [...parentPath, index]

      row.path = path
      row.depth = depth

      row.isExpanded =
        (row.original && row.original[expandedKey]) || getBy(expanded, path)

      if (paginateSubRows || (!paginateSubRows && row.depth === 0)) {
        expandedRows.push(row)
      }

      if (row.isExpanded && row.subRows && row.subRows.length) {
        row.subRows.forEach((row, i) => handleRow(row, i, depth + 1, path))
      }

      return row
    }

    rows.forEach((row, i) => handleRow(row, i))

    return expandedRows
  }, [debug, rows, expandedKey, expanded, paginateSubRows])

  const expandedDepth = findExpandedDepth(expanded)

  return {
    ...props,
    toggleExpandedByPath,
    expandedDepth,
    rows: expandedRows
  }
}

function findExpandedDepth(obj, depth = 1) {
  return Object.values(obj).reduce((prev, curr) => {
    if (typeof curr === 'object') {
      return Math.max(prev, findExpandedDepth(curr, depth + 1))
    }
    return depth
  }, 0)
}
