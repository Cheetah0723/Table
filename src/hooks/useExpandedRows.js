import { useMemo, useEffect } from 'react'

import { getBy } from '../utils'

export default function useExpandedRows ({
  debug,
  expanded,
  expandedKey,
  columns,
  rows,
  setState,
  actions,
}) {
  return useMemo(
    () => {
      if (debug) console.info('getExpandedRows')

      const expandedRows = []

      // Here we do some mutation, but it's the last stage in the
      // immutable process so this is safe
      const handleRow = (row, index, depth = 0, parentPath = []) => {
        // Compute some final state for the row
        const path = [...parentPath, index]

        row.path = path
        row.depth = depth

        row.isExpanded = (row.original && row.original[expandedKey]) || getBy(expanded, path)

        row.cells = columns.map(column => {
          const cell = {
            column,
            row,
            state: null,
            value: row.values[column.id],
          }

          return cell
        })

        expandedRows.push(row)

        if (row.isExpanded && row.subRows && row.subRows.length) {
          row.subRows.forEach((row, i) => handleRow(row, i, depth + 1, path))
        }
      }

      rows.forEach((row, i) => handleRow(row, i))

      return expandedRows
    },
    [rows, expanded, columns]
  )
}
