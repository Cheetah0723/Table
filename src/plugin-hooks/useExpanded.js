import React from 'react'

import {
  actions,
  makePropGetter,
  expandRows,
  useMountedLayoutEffect,
  useGetLatest,
} from '../utils'
import { useConsumeHookGetter, functionalUpdate } from '../publicUtils'

// Actions
actions.toggleExpanded = 'toggleExpanded'
actions.toggleAllExpanded = 'toggleAllExpanded'
actions.setExpanded = 'setExpanded'
actions.resetExpanded = 'resetExpanded'

export const useExpanded = hooks => {
  hooks.getExpandedToggleProps = [defaultGetExpandedToggleProps]
  hooks.stateReducers.push(reducer)
  hooks.useInstance.push(useInstance)
}

useExpanded.pluginName = 'useExpanded'

const defaultGetExpandedToggleProps = (props, { row }) => [
  props,
  {
    onClick: e => {
      e.persist()
      row.toggleExpanded()
    },
    style: {
      cursor: 'pointer',
    },
    title: 'Toggle Expanded',
  },
]

// Reducer
function reducer(state, action, previousState, instance) {
  if (action.type === actions.init) {
    return {
      expanded: {},
      ...state,
    }
  }

  if (action.type === actions.resetExpanded) {
    return {
      ...state,
      expanded: instance.initialState.expanded || {},
    }
  }

  if (action.type === actions.setExpanded) {
    return {
      ...state,
      expanded: functionalUpdate(action.expanded, state.expanded),
    }
  }

  if (action.type === actions.toggleExpanded) {
    const { id, expanded: setExpanded } = action
    const exists = state.expanded[id]

    const shouldExist =
      typeof setExpanded !== 'undefined' ? setExpanded : !exists

    if (!exists && shouldExist) {
      return {
        ...state,
        expanded: {
          ...state.expanded,
          [id]: true,
        },
      }
    } else if (exists && !shouldExist) {
      const { [id]: _, ...rest } = state.expanded
      return {
        ...state,
        expanded: rest,
      }
    } else {
      return state
    }
  }
}

function useInstance(instance) {
  const {
    data,
    rows,
    manualExpandedKey = 'expanded',
    paginateExpandedRows = true,
    expandSubRows = true,
    hooks,
    autoResetExpanded = true,
    state: { expanded },
    dispatch,
  } = instance

  const getAutoResetExpanded = useGetLatest(autoResetExpanded)

  // Bypass any effects from firing when this changes
  useMountedLayoutEffect(() => {
    if (getAutoResetExpanded()) {
      dispatch({ type: actions.resetExpanded })
    }
  }, [dispatch, data])

  const toggleExpanded = (id, expanded) => {
    dispatch({ type: actions.toggleExpanded, id, expanded })
  }

  // use reference to avoid memory leak in #1608
  const getInstance = useGetLatest(instance)

  const getExpandedTogglePropsHooks = useConsumeHookGetter(
    getInstance().hooks,
    'getExpandedToggleProps'
  )

  hooks.prepareRow.push(row => {
    row.toggleExpanded = set => instance.toggleExpanded(row.id, set)

    row.getExpandedToggleProps = makePropGetter(getExpandedTogglePropsHooks(), {
      instance: getInstance(),
      row,
    })
  })

  const expandedRows = React.useMemo(() => {
    if (paginateExpandedRows) {
      return expandRows(rows, { manualExpandedKey, expanded, expandSubRows })
    }

    return rows
  }, [paginateExpandedRows, rows, manualExpandedKey, expanded, expandSubRows])

  const expandedDepth = React.useMemo(() => findExpandedDepth(expanded), [
    expanded,
  ])

  Object.assign(instance, {
    preExpandedRows: rows,
    expandedRows,
    rows: expandedRows,
    toggleExpanded,
    expandedDepth,
  })
}

function findExpandedDepth(expanded) {
  let maxDepth = 0

  Object.keys(expanded).forEach(id => {
    const splitId = id.split('.')
    maxDepth = Math.max(maxDepth, splitId.length)
  })

  return maxDepth
}
