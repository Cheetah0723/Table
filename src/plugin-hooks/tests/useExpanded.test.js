import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { useTable } from '../../hooks/useTable'
import { useExpanded } from '../useExpanded'

const makeData = () => [
  {
    firstName: 'tanner',
    lastName: 'linsley',
    age: 29,
    visits: 100,
    status: 'In Relationship',
    progress: 50,
  },
  {
    firstName: 'derek',
    lastName: 'perkins',
    age: 40,
    visits: 40,
    status: 'Single',
    progress: 80,
  },
  {
    firstName: 'joe',
    lastName: 'bergevin',
    age: 45,
    visits: 20,
    status: 'Complicated',
    progress: 10,
  },
  {
    firstName: 'jaylen',
    lastName: 'linsley',
    age: 26,
    visits: 99,
    status: 'In Relationship',
    progress: 70,
  },
]

const data = makeData()

data[0].subRows = makeData()
data[0].subRows[0].subRows = makeData()
data[0].subRows[0].subRows[0].subRows = makeData()

function Table({ columns: userColumns, data, SubComponent }) {
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
    flatColumns,
    state: { expanded },
  } = useTable(
    {
      columns: userColumns,
      data,
    },
    useExpanded
  )

  return (
    <>
      <pre>
        <code>{JSON.stringify({ expanded: expanded }, null, 2)}</code>
      </pre>
      <table {...getTableProps()}>
        <thead>
          {headerGroups.map(headerGroup => (
            <tr {...headerGroup.getHeaderGroupProps()}>
              {headerGroup.headers.map(column => (
                <th {...column.getHeaderProps()}>{column.render('Header')}</th>
              ))}
            </tr>
          ))}
        </thead>
        <tbody {...getTableBodyProps()}>
          {rows.map((row, i) => {
            prepareRow(row)
            const { key, ...rowProps } = row.getRowProps()
            return (
              <React.Fragment key={key}>
                <tr {...rowProps}>
                  {row.cells.map(cell => {
                    return (
                      <td {...cell.getCellProps()}>{cell.render('Cell')}</td>
                    )
                  })}
                </tr>
                {!row.subRows.length && row.isExpanded ? (
                  <tr>
                    <td colSpan={flatColumns.length}>
                      {SubComponent({ row })}
                    </td>
                  </tr>
                ) : null}
              </React.Fragment>
            )
          })}
        </tbody>
      </table>
    </>
  )
}

function App() {
  const columns = React.useMemo(
    () => [
      {
        Header: () => null,
        id: 'expander',
        Cell: ({ row }) => (
          <span
            style={{
              cursor: 'pointer',
              paddingLeft: `${row.depth * 2}rem`,
            }}
            onClick={() => row.toggleExpanded()}
          >
            {row.isExpanded ? 'ðŸ‘‡' : 'ðŸ‘‰'}
          </span>
        ),
      },
      {
        Header: 'Name',
        columns: [
          {
            Header: 'First Name',
            accessor: 'firstName',
          },
          {
            Header: 'Last Name',
            accessor: 'lastName',
          },
        ],
      },
      {
        Header: 'Info',
        columns: [
          {
            Header: 'Age',
            accessor: 'age',
          },
          {
            Header: 'Visits',
            accessor: 'visits',
          },
          {
            Header: 'Status',
            accessor: 'status',
          },
          {
            Header: 'Profile Progress',
            accessor: 'progress',
          },
        ],
      },
    ],
    []
  )

  return (
    <Table
      columns={columns}
      data={data}
      SubComponent={({ row }) => (
        <pre>
          <code>{JSON.stringify({ values: row.values }, null, 2)}</code>
        </pre>
      )}
    />
  )
}

test('renders an expandable table', () => {
  const { getAllByText, asFragment } = render(<App />)

  let expandButtons = getAllByText('ðŸ‘‰')

  const before = asFragment()

  fireEvent.click(expandButtons[0])

  const after1 = asFragment()

  expandButtons = getAllByText('ðŸ‘‰')
  fireEvent.click(expandButtons[0])

  const after2 = asFragment()

  expandButtons = getAllByText('ðŸ‘‰')
  fireEvent.click(expandButtons[0])

  const after3 = asFragment()

  expandButtons = getAllByText('ðŸ‘‰')
  fireEvent.click(expandButtons[0])

  const after4 = asFragment()

  expandButtons = getAllByText('ðŸ‘‡')
  expandButtons.reverse().forEach(button => {
    fireEvent.click(button)
  })

  const after5 = asFragment()

  expect(before).toMatchDiffSnapshot(after1)
  expect(after1).toMatchDiffSnapshot(after2)
  expect(after2).toMatchDiffSnapshot(after3)
  expect(after3).toMatchDiffSnapshot(after4)
  expect(after4).toMatchDiffSnapshot(after5)
})
