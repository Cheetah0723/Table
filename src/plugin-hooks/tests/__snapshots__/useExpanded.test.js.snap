// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an expandable table 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -1,10 +1,12 @@
  <DocumentFragment>
    <pre>
      <code>
        {
-   "expanded": []
+   "expanded": [
+     "0"
+   ]
  }
      </code>
    </pre>
    <table
      class=""
@@ -86,10 +88,53 @@
          <td
            class=""
          >
            <span
              style="cursor: pointer; padding-left: 0rem;"
+           >
+             ðŸ‘‡
+           </span>
+         </td>
+         <td
+           class=""
+         >
+           tanner
+         </td>
+         <td
+           class=""
+         >
+           linsley
+         </td>
+         <td
+           class=""
+         >
+           29
+         </td>
+         <td
+           class=""
+         >
+           100
+         </td>
+         <td
+           class=""
+         >
+           In Relationship
+         </td>
+         <td
+           class=""
+         >
+           50
+         </td>
+       </tr>
+       <tr
+         class=""
+       >
+         <td
+           class=""
+         >
+           <span
+             style="cursor: pointer; padding-left: 2rem;"
            >
              ðŸ‘‰
            </span>
          </td>
          <td
@@ -119,10 +164,139 @@
          </td>
          <td
            class=""
          >
            50
+         </td>
+       </tr>
+       <tr
+         class=""
+       >
+         <td
+           class=""
+         >
+           <span
+             style="cursor: pointer; padding-left: 2rem;"
+           >
+             ðŸ‘‰
+           </span>
+         </td>
+         <td
+           class=""
+         >
+           derek
+         </td>
+         <td
+           class=""
+         >
+           perkins
+         </td>
+         <td
+           class=""
+         >
+           40
+         </td>
+         <td
+           class=""
+         >
+           40
+         </td>
+         <td
+           class=""
+         >
+           Single
+         </td>
+         <td
+           class=""
+         >
+           80
+         </td>
+       </tr>
+       <tr
+         class=""
+       >
+         <td
+           class=""
+         >
+           <span
+             style="cursor: pointer; padding-left: 2rem;"
+           >
+             ðŸ‘‰
+           </span>
+         </td>
+         <td
+           class=""
+         >
+           joe
+         </td>
+         <td
+           class=""
+         >
+           bergevin
+         </td>
+         <td
+           class=""
+         >
+           45
+         </td>
+         <td
+           class=""
+         >
+           20
+         </td>
+         <td
+           class=""
+         >
+           Complicated
+         </td>
+         <td
+           class=""
+         >
+           10
+         </td>
+       </tr>
+       <tr
+         class=""
+       >
+         <td
+           class=""
+         >
+           <span
+             style="cursor: pointer; padding-left: 2rem;"
+           >
+             ðŸ‘‰
+           </span>
+         </td>
+         <td
+           class=""
+         >
+           jaylen
+         </td>
+         <td
+           class=""
+         >
+           linsley
+         </td>
+         <td
+           class=""
+         >
+           26
+         </td>
+         <td
+           class=""
+         >
+           99
+         </td>
+         <td
+           class=""
+         >
+           In Relationship
+         </td>
+         <td
+           class=""
+         >
+           70
          </td>
        </tr>
        <tr
          class=""
        >
`;

exports[`renders an expandable table 2`] = `
Snapshot Diff:
- First value
+ Second value

@@ -1,11 +1,12 @@
  <DocumentFragment>
    <pre>
      <code>
        {
    "expanded": [
-     "0"
+     "0",
+     "0.0"
    ]
  }
      </code>
    </pre>
    <table
@@ -131,11 +132,54 @@
          <td
            class=""
          >
            <span
              style="cursor: pointer; padding-left: 2rem;"
+           >
+             ðŸ‘‡
+           </span>
+         </td>
+         <td
+           class=""
+         >
+           tanner
+         </td>
+         <td
+           class=""
+         >
+           linsley
+         </td>
+         <td
+           class=""
+         >
+           29
+         </td>
+         <td
+           class=""
+         >
+           100
+         </td>
+         <td
+           class=""
+         >
+           In Relationship
+         </td>
+         <td
+           class=""
+         >
+           50
+         </td>
+       </tr>
+       <tr
+         class=""
+       >
+         <td
+           class=""
          >
+           <span
+             style="cursor: pointer; padding-left: 4rem;"
+           >
              ðŸ‘‰
            </span>
          </td>
          <td
            class=""
@@ -164,10 +208,139 @@
          </td>
          <td
            class=""
          >
            50
+         </td>
+       </tr>
+       <tr
+         class=""
+       >
+         <td
+           class=""
+         >
+           <span
+             style="cursor: pointer; padding-left: 4rem;"
+           >
+             ðŸ‘‰
+           </span>
+         </td>
+         <td
+           class=""
+         >
+           derek
+         </td>
+         <td
+           class=""
+         >
+           perkins
+         </td>
+         <td
+           class=""
+         >
+           40
+         </td>
+         <td
+           class=""
+         >
+           40
+         </td>
+         <td
+           class=""
+         >
+           Single
+         </td>
+         <td
+           class=""
+         >
+           80
+         </td>
+       </tr>
+       <tr
+         class=""
+       >
+         <td
+           class=""
+         >
+           <span
+             style="cursor: pointer; padding-left: 4rem;"
+           >
+             ðŸ‘‰
+           </span>
+         </td>
+         <td
+           class=""
+         >
+           joe
+         </td>
+         <td
+           class=""
+         >
+           bergevin
+         </td>
+         <td
+           class=""
+         >
+           45
+         </td>
+         <td
+           class=""
+         >
+           20
+         </td>
+         <td
+           class=""
+         >
+           Complicated
+         </td>
+         <td
+           class=""
+         >
+           10
+         </td>
+       </tr>
+       <tr
+         class=""
+       >
+         <td
+           class=""
+         >
+           <span
+             style="cursor: pointer; padding-left: 4rem;"
+           >
+             ðŸ‘‰
+           </span>
+         </td>
+         <td
+           class=""
+         >
+           jaylen
+         </td>
+         <td
+           class=""
+         >
+           linsley
+         </td>
+         <td
+           class=""
+         >
+           26
+         </td>
+         <td
+           class=""
+         >
+           99
+         </td>
+         <td
+           class=""
+         >
+           In Relationship
+         </td>
+         <td
+           class=""
+         >
+           70
          </td>
        </tr>
        <tr
          class=""
        >
`;

exports[`renders an expandable table 3`] = `
Snapshot Diff:
- First value
+ Second value

@@ -2,11 +2,12 @@
    <pre>
      <code>
        {
    "expanded": [
      "0",
-     "0.0"
+     "0.0",
+     "0.0.0"
    ]
  }
      </code>
    </pre>
    <table
@@ -175,11 +176,54 @@
          <td
            class=""
          >
            <span
              style="cursor: pointer; padding-left: 4rem;"
+           >
+             ðŸ‘‡
+           </span>
+         </td>
+         <td
+           class=""
+         >
+           tanner
+         </td>
+         <td
+           class=""
+         >
+           linsley
+         </td>
+         <td
+           class=""
+         >
+           29
+         </td>
+         <td
+           class=""
+         >
+           100
+         </td>
+         <td
+           class=""
+         >
+           In Relationship
+         </td>
+         <td
+           class=""
+         >
+           50
+         </td>
+       </tr>
+       <tr
+         class=""
+       >
+         <td
+           class=""
          >
+           <span
+             style="cursor: pointer; padding-left: 6rem;"
+           >
              ðŸ‘‰
            </span>
          </td>
          <td
            class=""
@@ -208,10 +252,139 @@
          </td>
          <td
            class=""
          >
            50
+         </td>
+       </tr>
+       <tr
+         class=""
+       >
+         <td
+           class=""
+         >
+           <span
+             style="cursor: pointer; padding-left: 6rem;"
+           >
+             ðŸ‘‰
+           </span>
+         </td>
+         <td
+           class=""
+         >
+           derek
+         </td>
+         <td
+           class=""
+         >
+           perkins
+         </td>
+         <td
+           class=""
+         >
+           40
+         </td>
+         <td
+           class=""
+         >
+           40
+         </td>
+         <td
+           class=""
+         >
+           Single
+         </td>
+         <td
+           class=""
+         >
+           80
+         </td>
+       </tr>
+       <tr
+         class=""
+       >
+         <td
+           class=""
+         >
+           <span
+             style="cursor: pointer; padding-left: 6rem;"
+           >
+             ðŸ‘‰
+           </span>
+         </td>
+         <td
+           class=""
+         >
+           joe
+         </td>
+         <td
+           class=""
+         >
+           bergevin
+         </td>
+         <td
+           class=""
+         >
+           45
+         </td>
+         <td
+           class=""
+         >
+           20
+         </td>
+         <td
+           class=""
+         >
+           Complicated
+         </td>
+         <td
+           class=""
+         >
+           10
+         </td>
+       </tr>
+       <tr
+         class=""
+       >
+         <td
+           class=""
+         >
+           <span
+             style="cursor: pointer; padding-left: 6rem;"
+           >
+             ðŸ‘‰
+           </span>
+         </td>
+         <td
+           class=""
+         >
+           jaylen
+         </td>
+         <td
+           class=""
+         >
+           linsley
+         </td>
+         <td
+           class=""
+         >
+           26
+         </td>
+         <td
+           class=""
+         >
+           99
+         </td>
+         <td
+           class=""
+         >
+           In Relationship
+         </td>
+         <td
+           class=""
+         >
+           70
          </td>
        </tr>
        <tr
          class=""
        >
`;

exports[`renders an expandable table 4`] = `
Snapshot Diff:
- First value
+ Second value

@@ -3,11 +3,12 @@
      <code>
        {
    "expanded": [
      "0",
      "0.0",
-     "0.0.0"
+     "0.0.0",
+     "0.0.0.0"
    ]
  }
      </code>
    </pre>
    <table
@@ -220,11 +221,11 @@
            class=""
          >
            <span
              style="cursor: pointer; padding-left: 6rem;"
            >
-             ðŸ‘‰
+             ðŸ‘‡
            </span>
          </td>
          <td
            class=""
          >
@@ -252,10 +253,30 @@
          </td>
          <td
            class=""
          >
            50
+         </td>
+       </tr>
+       <tr>
+         <td
+           colspan="7"
+         >
+           <pre>
+             <code>
+               {
+   "values": {
+     "firstName": "tanner",
+     "lastName": "linsley",
+     "age": 29,
+     "visits": 100,
+     "status": "In Relationship",
+     "progress": 50
+   }
+ }
+             </code>
+           </pre>
          </td>
        </tr>
        <tr
          class=""
        >
`;

exports[`renders an expandable table 5`] = `
Snapshot Diff:
- First value
+ Second value

@@ -1,15 +1,10 @@
  <DocumentFragment>
    <pre>
      <code>
        {
-   "expanded": [
-     "0",
-     "0.0",
-     "0.0.0",
-     "0.0.0.0"
-   ]
+   "expanded": []
  }
      </code>
    </pre>
    <table
      class=""
@@ -92,140 +87,11 @@
            class=""
          >
            <span
              style="cursor: pointer; padding-left: 0rem;"
            >
-             ðŸ‘‡
-           </span>
-         </td>
-         <td
-           class=""
-         >
-           tanner
-         </td>
-         <td
-           class=""
-         >
-           linsley
-         </td>
-         <td
-           class=""
-         >
-           29
-         </td>
-         <td
-           class=""
-         >
-           100
-         </td>
-         <td
-           class=""
-         >
-           In Relationship
-         </td>
-         <td
-           class=""
-         >
-           50
-         </td>
-       </tr>
-       <tr
-         class=""
-       >
-         <td
-           class=""
-         >
-           <span
-             style="cursor: pointer; padding-left: 2rem;"
-           >
-             ðŸ‘‡
-           </span>
-         </td>
-         <td
-           class=""
-         >
-           tanner
-         </td>
-         <td
-           class=""
-         >
-           linsley
-         </td>
-         <td
-           class=""
-         >
-           29
-         </td>
-         <td
-           class=""
-         >
-           100
-         </td>
-         <td
-           class=""
-         >
-           In Relationship
-         </td>
-         <td
-           class=""
-         >
-           50
-         </td>
-       </tr>
-       <tr
-         class=""
-       >
-         <td
-           class=""
-         >
-           <span
-             style="cursor: pointer; padding-left: 4rem;"
-           >
-             ðŸ‘‡
-           </span>
-         </td>
-         <td
-           class=""
-         >
-           tanner
-         </td>
-         <td
-           class=""
-         >
-           linsley
-         </td>
-         <td
-           class=""
-         >
-           29
-         </td>
-         <td
-           class=""
-         >
-           100
-         </td>
-         <td
-           class=""
-         >
-           In Relationship
-         </td>
-         <td
-           class=""
-         >
-           50
-         </td>
-       </tr>
-       <tr
-         class=""
-       >
-         <td
-           class=""
-         >
-           <span
-             style="cursor: pointer; padding-left: 6rem;"
-           >
-             ðŸ‘‡
+             ðŸ‘‰
            </span>
          </td>
          <td
            class=""
          >
@@ -253,417 +119,10 @@
          </td>
          <td
            class=""
          >
            50
-         </td>
-       </tr>
-       <tr>
-         <td
-           colspan="7"
-         >
-           <pre>
-             <code>
-               {
-   "values": {
-     "firstName": "tanner",
-     "lastName": "linsley",
-     "age": 29,
-     "visits": 100,
-     "status": "In Relationship",
-     "progress": 50
-   }
- }
-             </code>
-           </pre>
-         </td>
-       </tr>
-       <tr
-         class=""
-       >
-         <td
-           class=""
-         >
-           <span
-             style="cursor: pointer; padding-left: 6rem;"
-           >
-             ðŸ‘‰
-           </span>
-         </td>
-         <td
-           class=""
-         >
-           derek
-         </td>
-         <td
-           class=""
-         >
-           perkins
-         </td>
-         <td
-           class=""
-         >
-           40
-         </td>
-         <td
-           class=""
-         >
-           40
-         </td>
-         <td
-           class=""
-         >
-           Single
-         </td>
-         <td
-           class=""
-         >
-           80
-         </td>
-       </tr>
-       <tr
-         class=""
-       >
-         <td
-           class=""
-         >
-           <span
-             style="cursor: pointer; padding-left: 6rem;"
-           >
-             ðŸ‘‰
-           </span>
-         </td>
-         <td
-           class=""
-         >
-           joe
-         </td>
-         <td
-           class=""
-         >
-           bergevin
-         </td>
-         <td
-           class=""
-         >
-           45
-         </td>
-         <td
-           class=""
-         >
-           20
-         </td>
-         <td
-           class=""
-         >
-           Complicated
-         </td>
-         <td
-           class=""
-         >
-           10
-         </td>
-       </tr>
-       <tr
-         class=""
-       >
-         <td
-           class=""
-         >
-           <span
-             style="cursor: pointer; padding-left: 6rem;"
-           >
-             ðŸ‘‰
-           </span>
-         </td>
-         <td
-           class=""
-         >
-           jaylen
-         </td>
-         <td
-           class=""
-         >
-           linsley
-         </td>
-         <td
-           class=""
-         >
-           26
-         </td>
-         <td
-           class=""
-         >
-           99
-         </td>
-         <td
-           class=""
-         >
-           In Relationship
-         </td>
-         <td
-           class=""
-         >
-           70
-         </td>
-       </tr>
-       <tr
-         class=""
-       >
-         <td
-           class=""
-         >
-           <span
-             style="cursor: pointer; padding-left: 4rem;"
-           >
-             ðŸ‘‰
-           </span>
-         </td>
-         <td
-           class=""
-         >
-           derek
-         </td>
-         <td
-           class=""
-         >
-           perkins
-         </td>
-         <td
-           class=""
-         >
-           40
-         </td>
-         <td
-           class=""
-         >
-           40
-         </td>
-         <td
-           class=""
-         >
-           Single
-         </td>
-         <td
-           class=""
-         >
-           80
-         </td>
-       </tr>
-       <tr
-         class=""
-       >
-         <td
-           class=""
-         >
-           <span
-             style="cursor: pointer; padding-left: 4rem;"
-           >
-             ðŸ‘‰
-           </span>
-         </td>
-         <td
-           class=""
-         >
-           joe
-         </td>
-         <td
-           class=""
-         >
-           bergevin
-         </td>
-         <td
-           class=""
-         >
-           45
-         </td>
-         <td
-           class=""
-         >
-           20
-         </td>
-         <td
-           class=""
-         >
-           Complicated
-         </td>
-         <td
-           class=""
-         >
-           10
-         </td>
-       </tr>
-       <tr
-         class=""
-       >
-         <td
-           class=""
-         >
-           <span
-             style="cursor: pointer; padding-left: 4rem;"
-           >
-             ðŸ‘‰
-           </span>
-         </td>
-         <td
-           class=""
-         >
-           jaylen
-         </td>
-         <td
-           class=""
-         >
-           linsley
-         </td>
-         <td
-           class=""
-         >
-           26
-         </td>
-         <td
-           class=""
-         >
-           99
-         </td>
-         <td
-           class=""
-         >
-           In Relationship
-         </td>
-         <td
-           class=""
-         >
-           70
-         </td>
-       </tr>
-       <tr
-         class=""
-       >
-         <td
-           class=""
-         >
-           <span
-             style="cursor: pointer; padding-left: 2rem;"
-           >
-             ðŸ‘‰
-           </span>
-         </td>
-         <td
-           class=""
-         >
-           derek
-         </td>
-         <td
-           class=""
-         >
-           perkins
-         </td>
-         <td
-           class=""
-         >
-           40
-         </td>
-         <td
-           class=""
-         >
-           40
-         </td>
-         <td
-           class=""
-         >
-           Single
-         </td>
-         <td
-           class=""
-         >
-           80
-         </td>
-       </tr>
-       <tr
-         class=""
-       >
-         <td
-           class=""
-         >
-           <span
-             style="cursor: pointer; padding-left: 2rem;"
-           >
-             ðŸ‘‰
-           </span>
-         </td>
-         <td
-           class=""
-         >
-           joe
-         </td>
-         <td
-           class=""
-         >
-           bergevin
-         </td>
-         <td
-           class=""
-         >
-           45
-         </td>
-         <td
-           class=""
-         >
-           20
-         </td>
-         <td
-           class=""
-         >
-           Complicated
-         </td>
-         <td
-           class=""
-         >
-           10
-         </td>
-       </tr>
-       <tr
-         class=""
-       >
-         <td
-           class=""
-         >
-           <span
-             style="cursor: pointer; padding-left: 2rem;"
-           >
-             ðŸ‘‰
-           </span>
-         </td>
-         <td
-           class=""
-         >
-           jaylen
-         </td>
-         <td
-           class=""
-         >
-           linsley
-         </td>
-         <td
-           class=""
-         >
-           26
-         </td>
-         <td
-           class=""
-         >
-           99
-         </td>
-         <td
-           class=""
-         >
-           In Relationship
-         </td>
-         <td
-           class=""
-         >
-           70
          </td>
        </tr>
        <tr
          class=""
        >
`;
